File: src/ovc.ts
// Servicio OVC: la API oficial, la que se bloquea con volumen. Un 403 aislado no
// significa nada (lo da también una RC inventada); diagnóstico solo con catastro_estado.
import { pedirPermiso, registrarExito, registrarFallo } from "./limiter.js";

export const OVC_HOST = "ovc.catastro.meh.es";
const BASE = `http://${OVC_HOST}/ovcservweb/OVCSWLocalizacionRC`;

export const canario: { rc: string | null; provincia: string | null; municipio: string | null } =
  { rc: null, provincia: null, municipio: null };

export class OVCError extends Error {}

async function get(path: string, params: Record<string, string>): Promise<string> {
  await pedirPermiso(OVC_HOST);
  let texto: string;
  try {
    const r = await fetch(`${BASE}/${path}?` + new URLSearchParams(params));
    if (!r.ok) throw new OVCError(`HTTP ${r.status}`);
    texto = await r.text();
  } catch (e) {
    registrarFallo(OVC_HOST);
    throw e;
  }
  registrarExito(OVC_HOST);
  return texto;
}

function tag(xml: string, nombre: string): string | null {
  const m = new RegExp(`<(?:\\w+:)?${nombre}[^>]*>([^<]*)<`).exec(xml);
  return m?.[1]?.trim() || null;
}

function errores(xml: string): string | null {
  const m = /<(?:\w+:)?err[^>]*>([\s\S]*?)<\/(?:\w+:)?err>/.exec(xml);
  return m ? m[1].replace(/<[^>]+>/g, " ").replace(/\s+/g, " ").trim() : null;
}

function parsearDNP(xml: string) {
  const err = errores(xml);
  if (err) throw new OVCError(err);
  const pc1 = tag(xml, "pc1"), pc2 = tag(xml, "pc2");
  const car = tag(xml, "car"), cc1 = tag(xml, "cc1"), cc2 = tag(xml, "cc2");
  return {
    rc14: pc1 && pc2 ? pc1 + pc2 : null,
    rc20: pc1 && pc2 && car && cc1 && cc2 ? pc1 + pc2 + car + cc1 + cc2 : null,
    direccion: tag(xml, "ldt"),
    uso: tag(xml, "luso"),
    superficie_m2: tag(xml, "ssp") ?? tag(xml, "sfc"),
    paraje: tag(xml, "npa"),
    poligono: tag(xml, "cpo"),
    parcela: tag(xml, "cpa"),
  };
}

export async function consultaDNPRC(provincia: string, municipio: string, rc: string) {
  return parsearDNP(await get("OVCCallejero.asmx/Consulta_DNPRC",
    { Provincia: provincia, Municipio: municipio, RC: rc }));
}

export async function consultaDNPPP(provincia: string, municipio: string, poligono: string, parcela: string) {
  return parsearDNP(await get("OVCCallejero.asmx/Consulta_DNPPP",
    { Provincia: provincia, Municipio: municipio, Poligono: poligono, Parcela: parcela }));
}

export async function consultaRCCOOR(x: number, y: number, srs = "EPSG:25829") {
  const xml = await get("OVCCoordenadas.asmx/Consulta_RCCOOR",
    { SRS: srs, Coordenada_X: String(x), Coordenada_Y: String(y) });
  const err = errores(xml);
  if (err) throw new OVCError(err);
  const pc1 = tag(xml, "pc1"), pc2 = tag(xml, "pc2");
  return { rc14: pc1 && pc2 ? pc1 + pc2 : null, direccion: tag(xml, "ldt") };
}

export async function probar(rcOk: string | null, provincia: string, municipio: string) {
  const resultados: Record<string, string> = {};
  const casos: Array<[string, string | null]> = [["valida", rcOk], ["inventada", "99999Z999999ZZ"]];
  for (const [etiqueta, rc] of casos) {
    if (!rc) { resultados[etiqueta] = "sin_referencia"; continue; }
    try {
      await consultaDNPRC(provincia, municipio, rc);
      resultados[etiqueta] = "ok";
    } catch (e) {
      resultados[etiqueta] = e instanceof OVCError
        ? `error_datos: ${e.message}`
        : `error_red: ${(e as Error).message}`;
    }
  }
  return resultados;
}

File: src/catastro_mcp/ovc.py
"""Servicio OVC (ovc.catastro.meh.es): la API oficial, la que se bloquea con volumen.

Todas las llamadas pasan por el limitador. Un 403 aislado no significa nada:
lo devuelve también una RC inventada. Diagnóstico solo con el control de tres puntas
(catastro_estado)."""
import xml.etree.ElementTree as ET

import httpx

from .limiter import LIMITADOR

HOST = "ovc.catastro.meh.es"
BASE = f"http://{HOST}/ovcservweb"

# Referencia canario: una RC que sabemos que funciona (se fija tras la primera
# consulta con éxito; configurable).
CANARIO_RC = {"rc": None, "provincia": None, "municipio": None}


class OVCError(Exception):
    pass


def _get(path: str, params: dict) -> ET.Element:
    LIMITADOR.pedir_permiso(HOST)
    try:
        r = httpx.get(f"{BASE}/{path}", params=params, timeout=30, follow_redirects=True)
        r.raise_for_status()
    except Exception:
        LIMITADOR.registrar_fallo(HOST)
        raise
    LIMITADOR.registrar_exito(HOST)
    texto = r.text
    # quitar namespaces para parsear cómodo
    import re
    texto = re.sub(r'\sxmlns(:\w+)?="[^"]+"', "", texto, count=10)
    texto = re.sub(r"<(/?)\w+:", r"<\1", texto)
    return ET.fromstring(texto)


def _texto(root: ET.Element, tag: str) -> str | None:
    el = root.find(f".//{tag}")
    return el.text if el is not None and el.text else None


def _errores(root: ET.Element) -> str | None:
    err = root.find(".//err")
    if err is not None:
        return " ".join(x.text or "" for x in err.iter() if x.text)
    return None


def consulta_dnprc(provincia: str, municipio: str, rc: str) -> dict:
    root = _get(
        "OVCSWLocalizacionRC/OVCCallejero.asmx/Consulta_DNPRC",
        {"Provincia": provincia, "Municipio": municipio, "RC": rc},
    )
    return _parsear_dnp(root)


def consulta_dnppp(provincia: str, municipio: str, poligono: str, parcela: str) -> dict:
    root = _get(
        "OVCSWLocalizacionRC/OVCCallejero.asmx/Consulta_DNPPP",
        {"Provincia": provincia, "Municipio": municipio, "Poligono": poligono, "Parcela": parcela},
    )
    return _parsear_dnp(root)


def consulta_rccoor(x: float, y: float, srs: str = "EPSG:25829") -> dict:
    root = _get(
        "OVCSWLocalizacionRC/OVCCoordenadas.asmx/Consulta_RCCOOR",
        {"SRS": srs, "Coordenada_X": x, "Coordenada_Y": y},
    )
    err = _errores(root)
    if err:
        raise OVCError(err)
    pc1, pc2 = _texto(root, "pc1"), _texto(root, "pc2")
    return {
        "rc14": (pc1 or "") + (pc2 or "") or None,
        "direccion": _texto(root, "ldt"),
    }


def _parsear_dnp(root: ET.Element) -> dict:
    err = _errores(root)
    if err:
        raise OVCError(err)
    pc1, pc2 = _texto(root, "pc1"), _texto(root, "pc2")
    car, cc1, cc2 = _texto(root, "car"), _texto(root, "cc1"), _texto(root, "cc2")
    rc20 = None
    if pc1 and pc2 and car and cc1 and cc2:
        rc20 = pc1 + pc2 + car + cc1 + cc2
    return {
        "rc14": (pc1 or "") + (pc2 or "") or None,
        "rc20": rc20,
        "direccion": _texto(root, "ldt"),
        "uso": _texto(root, "luso"),
        "superficie_m2": _texto(root, "ssp") or _texto(root, "sfc"),
        "paraje": _texto(root, "npa"),
        "poligono": _texto(root, "cpp/cpo") or _texto(root, "cpo"),
        "parcela": _texto(root, "cpp/cpa") or _texto(root, "cpa"),
    }


def probar(rc_ok: str | None, provincia: str, municipio: str) -> dict:
    """Control de tres puntas contra OVC: RC buena, RC inventada. Devuelve estado."""
    resultados = {}
    rc_falsa = "99999Z999999ZZ"
    for etiqueta, rc in (("valida", rc_ok), ("inventada", rc_falsa)):
        if rc is None:
            resultados[etiqueta] = "sin_referencia"
            continue
        try:
            consulta_dnprc(provincia, municipio, rc)
            resultados[etiqueta] = "ok"
        except OVCError as e:
            resultados[etiqueta] = f"error_datos: {e}"
        except Exception as e:
            resultados[etiqueta] = f"error_red: {type(e).__name__}: {e}"
    return resultados
